cmd/acraserver/acraserver.go

```diff
@@ -178,8 +188,96 @@ func main() {
 			}
 		}()
 	}
-	if *withZone && !*disableHTTPApi {
-		go server.StartCommands()
+
+	sigHandlerSIGTERM, err := cmd.NewSignalHandler([]os.Signal{os.Interrupt, syscall.SIGTERM})
+	ErrorSignalChannel = sigHandlerSIGTERM.GetChannel()
+	if err != nil {
+		log.WithError(err).Errorln("can't register SIGTERM handler")
+		os.Exit(1)
+	}
+	go sigHandlerSIGTERM.Register()
+	sigHandlerSIGTERM.AddCallback(func() {
+		log.Infoln("Incoming SIGTERM or SIGINT")
+		// Stop accepting new connections
+		server.StopListeners()
+		// Wait a maximum of N seconds for existing connections to finish
+		err := server.WaitWithTimeout(ACRASERVER_WAIT_TIMEOUT * time.Second)
+		if err == errors.New("timeout") {
+			log.Warningf("Server shutdown Timeout: %d active connections will be cut", server.ConnectionsCounter())
+			server.Close()
+			os.Exit(1)
+		}
+		server.Close()
+		log.Infof("Server graceful shutdown completed, bye PID: %v", os.Getpid())
+		os.Exit(0)
+	})
+
+	sigHandlerSIGHUP, err := cmd.NewSignalHandler([]os.Signal{syscall.SIGHUP})
+	RestartSignalsChannel = sigHandlerSIGHUP.GetChannel()
+	if err != nil {
+		log.WithError(err).Errorln("can't register SIGHUP handler")
+		os.Exit(1)
 	}
-	server.Start()
+	sigHandlerSIGHUP.AddCallback(func() {
+		log.Infoln("Incoming SIGHUP")
+
+		// Stop accepting requests
+		server.StopListeners()
+
+		// Get socket file descriptor to pass it to fork
+		var fdACRA, fdAPI uintptr
+		fdACRA, err = network.ListenerFileDescriptor(server.listenerACRA)
+		if err != nil {
+			log.Fatalln("Fail to get acra-socket file descriptor:", err)
+		}
+		if *withZone || *enableHTTPApi {
+			fdAPI, err = network.ListenerFileDescriptor(server.listenerAPI)
+			if err != nil {
+				log.Fatalln("Fail to get api-socket file descriptor:", err)
+			}
+		}
+
+		// Set env flag for forked process
+		os.Setenv(GRACEFUL_ENV, "true")
+		execSpec := &syscall.ProcAttr{
+			Env:   os.Environ(),
+			Files: []uintptr{os.Stdin.Fd(), os.Stdout.Fd(), os.Stderr.Fd(), fdACRA, fdAPI},
+		}
+		// Fork new process
+		var fork, err = syscall.ForkExec(os.Args[0], os.Args, execSpec)
+		if err != nil {
+			log.Fatalln("Fail to fork", err)
+		}
+		log.Infof("Server forked to PID: %v", fork)
+
+		// Wait a maximum of N seconds for existing connections to finish
+		err = server.WaitWithTimeout(ACRASERVER_WAIT_TIMEOUT * time.Second)
+		if err == errors.New("timeout") {
+			log.Warningf("Server shutdown Timeout: %d active connections will be cut", server.ConnectionsCounter())
+			os.Exit(0)
+		}
+		log.Infof("Server graceful restart completed, bye PID: %v", os.Getpid())
+
+		// Stop the old server, all the connections have been closed and the new one is running
+		os.Exit(0)
+	})
+
+	if *withZone || *enableHTTPApi {
+		if os.Getenv(GRACEFUL_ENV) == "true" {
+			go server.StartCommandsFromFileDescriptor(DESCRIPTOR_API)
+		} else {
+			go server.StartCommands()
+		}
+
+	}
+
+	log.Infof("PID: %v", os.Getpid())
+	if os.Getenv(GRACEFUL_ENV) == "true" {
```

Comment:
what about to merge this block with if block above (where used os.Getenv condition)?